test 30 failed

I've downloaded up-to-date sources and tried to make test it. The following problem was found:
```
RunTests: test 27: bs.vote.predict OK
RunTests: test 28: stderr OK
RunTests: test 28: bs.prvote.predict OK
RunTests: test 29: stderr OK
*** Error in `../vowpalwabbit/vw': double free or corruption (out): 0x00007fa2c40009b0 ***
Aborted
RunTests: test 30: '/usr/bin/timeout 30 ../vowpalwabbit/vw -d train-sets/0001.dat -f models/mask.model --invert_hash mask.predict --l1 0.01' failed (exitcode=134)
using l1 regularization = 0.01
final_regressor = models/mask.model
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = train-sets/0001.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
1.000000 1.000000            1            1.0   1.0000   0.0000       51
/usr/bin/timeout: the monitored command dumped core
RunTests: exiting with status=134
Makefile:125: recipe for target 'test' failed
make: *** [test] Error 134
```
while Travic-CI reports that all tests are fine.

My configuration is: Linux 4.2.0-23-generic, gcc version 5.2.1 20151010 (Ubuntu 5.2.1-22ubuntu2)
Will try to debug.

--- a/library/gd_mf_weights.cc
+++ b/library/gd_mf_weights.cc
@@ -83,9 +83,9 @@ int main(int argc, char *argv[])
     // write out features for left namespace
     features& left = ec->feature_space[left_ns];
     for (size_t i = 0; i < left.size(); ++i)
-      { left_linear << left.space_names[i].second << '\t' << weights[left.indicies[i] & mask];
+      { left_linear << left.space_names[i].get()->second << '\t' << weights[left.indicies[i] & mask];
 
-        left_quadratic << left.space_names[i].second;
+        left_quadratic << left.space_names[i].get()->second;
         for (size_t k = 1; k <= rank; k++)
           left_quadratic << '\t' << weights[(left.indicies[i] + k) & mask];
       }
@@ -95,9 +95,9 @@ int main(int argc, char *argv[])
     // write out features for right namespace
     features& right = ec->feature_space[right_ns];
     for (size_t i = 0; i < right.size(); ++i)
-      { right_linear << right.space_names[i].second << '\t' << weights[left.indicies[i] & mask];
+      { right_linear << right.space_names[i].get()->second << '\t' << weights[left.indicies[i] & mask];
 
-        right_quadratic << right.space_names[i].second;
+        right_quadratic << right.space_names[i].get()->second;
         for (size_t k = 1; k <= rank; k++)
           right_quadratic << '\t' << weights[(left.indicies[i] + k + rank) & mask];
       }

--- a/vowpalwabbit/feature_group.h
+++ b/vowpalwabbit/feature_group.h
@@ -1,6 +1,9 @@
+#include <memory>
+
 typedef float feature_value;
 typedef uint64_t feature_index;
 typedef pair<string, string> audit_strings;
+typedef std::shared_ptr<audit_strings> audit_strings_ptr;
 
 struct feature {//sparse feature definition for the library interface
   float x;
@@ -27,13 +30,13 @@ inline int order_features(const void* first, const void* second)
 struct features { // the core definition of a set of features.
   v_array<feature_value> values; // Always needed.
   v_array<feature_index> indicies; //Optional for dense data.
-  v_array<audit_strings> space_names; //Optional for audit mode.
+  v_array<audit_strings_ptr> space_names; //Optional for audit mode.
   float sum_feat_sq;
   features()
   {
     values = v_init<feature_value>();
     indicies = v_init<feature_index>();
-    space_names = v_init<audit_strings>();
+    space_names = v_init<audit_strings_ptr>();
     sum_feat_sq = 0.f;
   }
   inline size_t size() const { return values.size(); }
@@ -42,7 +45,7 @@ struct features { // the core definition of a set of features.
   void free_space_names(size_t i)
   {
 	  for (; i < space_names.size(); i++)
-		  space_names[i].~audit_strings();
+                  space_names[i].~audit_strings_ptr();
   }
 
   template<class R, void (*T)(R&, feature_value)> inline void foreach_feature(R& dat)
@@ -81,10 +84,10 @@ struct features { // the core definition of a set of features.
   {
     feature_value* vp = values.begin + j;
     feature_index* ip = indicies.begin + j;
-    audit_strings* ap = space_names.begin + j;
+    audit_strings_ptr* ap = space_names.begin + j;
     feature_value* end = values.begin + k;
     for (;vp != end; ++vp, ++ip, ++ap)
-      T(dat,*vp,*ip, ap);
+      T(dat,*vp,*ip, ap->get());
   }
 
   template<class R, void (*T)(R&, feature_value, feature_index, audit_strings*)> inline void foreach_feature(R& dat, size_t j)
@@ -134,7 +137,7 @@ struct features { // the core definition of a set of features.
       {
         feature_slice temp = {values[i], indicies[i] & parse_mask, audit_strings("", "")};
         if (space_names.size() != 0)
-          temp.space_name = space_names[i];
+          temp.space_name = *space_names[i].get();
         slice.push_back(temp);
       }
     qsort(slice.begin, slice.size(), sizeof(feature_slice), order_features);
@@ -143,7 +146,7 @@ struct features { // the core definition of a set of features.
         values[i] = slice[i].v;
         indicies[i] = slice[i].i;
         if (space_names.size() > 0)
-          space_names[i] = slice[i].space_name;
+          *space_names[i].get() = slice[i].space_name;
       }
     slice.delete_v();
     return true;

--- a/vowpalwabbit/gd.cc
+++ b/vowpalwabbit/gd.cc
@@ -252,7 +252,7 @@ void print_features(vw& all, example& ec)
       {
         features& fs = ec.feature_space[*i];
         for (size_t j = 0; j < fs.size(); ++j)
-          { cout << '\t' << fs.space_names[j].first << '^' << fs.space_names[j].second << ':' << ((fs.indicies[j] >> all.reg.stride_shift) & all.parse_mask) << ':' << fs.values[j];
+          { cout << '\t' << fs.space_names[j].get()->first << '^' << fs.space_names[j].get()->second << ':' << ((fs.indicies[j] >> all.reg.stride_shift) & all.parse_mask) << ':' << fs.values[j];
             for (size_t k = 0; k < all.lda; k++)
               cout << ':' << weights[(fs.indicies[j]+k) & all.reg.weight_mask];
           }
@@ -269,7 +269,7 @@ void print_features(vw& all, example& ec)
 	if (fs.space_names.size() > 0)
 	  for (size_t j = 0; j < fs.size(); ++j)
 	    {
-	      audit_interaction(dat, &fs.space_names[j]);
+          audit_interaction(dat, fs.space_names[j].get());
 	      audit_feature(dat, fs.values[j], (uint32_t)fs.indicies[j] + ec.ft_offset);
 	      audit_interaction(dat, NULL);
 	    }

--- a/vowpalwabbit/gd_mf.cc
+++ b/vowpalwabbit/gd_mf.cc
@@ -41,7 +41,7 @@ void mf_print_offset_features(gdmf& d, example& ec, size_t offset)
         {
           cout << '\t';
           if (audit)
-            cout << fs.space_names[j].first << '^' << fs.space_names[j].second << ':';
+            cout << fs.space_names[j].get()->first << '^' << fs.space_names[j].get()->second << ':';
           cout << fs.indicies[j] <<"(" << ((fs.indicies[j] + offset) & mask)  << ")" << ':' << fs.values[j];
           cout << ':' << weights[(fs.indicies[j] + offset) & mask];
         }
@@ -55,11 +55,11 @@ void mf_print_offset_features(gdmf& d, example& ec, size_t offset)
         features& fs2 = ec.feature_space[(int)(*i)[1]];
         for (size_t j = 0; j < fs1.size(); ++j)
           for (size_t l = 0; l < fs2.size(); ++l)
-          { cout << '\t' << fs1.space_names[j].first << k << '^' << fs1.space_names[j].second << ':' << ((fs1.indicies[j]+k)&mask)
+          { cout << '\t' << fs1.space_names[j].get()->first << k << '^' << fs1.space_names[j].get()->second << ':' << ((fs1.indicies[j]+k)&mask)
                  <<"(" << ((fs1.indicies[j] + offset +k) & mask)  << ")" << ':' << fs1.values[j];
             cout << ':' << weights[(fs1.indicies[j] + offset + k) & mask];
 
-            cout << ':' << fs2.space_names[l].first << k << '^' << fs2.space_names[l].second << ':' << ((fs2.indicies[l]+k+d.rank)&mask)
+            cout << ':' << fs2.space_names[l].get()->first << k << '^' << fs2.space_names[l].get()->second << ':' << ((fs2.indicies[l]+k+d.rank)&mask)
                  <<"(" << ((fs2.indicies[l] + offset +k+d.rank) & mask)  << ")" << ':' << fs2.values[l];
             cout << ':' << weights[(fs2.indicies[l] + offset + k+d.rank) & mask];
 

--- a/vowpalwabbit/interactions.h
+++ b/vowpalwabbit/interactions.h
@@ -163,7 +163,7 @@ template<class R>
 
                 for(size_t i = 0; i < first.indicies.size(); ++i)
                   { id.halfhash = FNV_prime * (uint64_t)first.indicies[i];
-                    if (audit) audit_func(dat, &first.space_names[i]);
+                    if (audit) audit_func(dat, first.space_names[i].get());
                     // next index differs for permutations and simple combinations
                     id.ft_value = first.values[i];
                     size_t j=0;
@@ -197,7 +197,7 @@ template<class R>
 
               for(size_t i = 0; i < first.indicies.size(); ++i)
               {
-                if(audit) audit_func(dat, &first.space_names[i]);
+                if(audit) audit_func(dat, first.space_names[i].get());
                 const uint64_t halfhash1 = FNV_prime * (uint64_t)first.indicies[i];
                 const float& ft_value = first.values[i];
                 size_t j=0;
@@ -206,7 +206,7 @@ template<class R>
 
                 for (; j < second.indicies.size(); ++j)
                 { //f3 x k*(f2 x k*f1)
-                  if(audit) audit_func(dat, &second.space_names[j]);
+                  if(audit) audit_func(dat, second.space_names[j].get());
                   id.halfhash = FNV_prime * (halfhash1 ^ (uint64_t)second.indicies[j]);
                   id.ft_value = INTERACTION_VALUE(ft_value, second.values[j]);
 
@@ -325,7 +325,7 @@ template<class R>
               else
                 next_data->loop_idx = 0;
 
-              if(audit) audit_func(dat, &fs.space_names[feature]);
+              if(audit) audit_func(dat, fs.space_names[feature].get());
 
               if (cur_data == fgd) // first namespace
                 { next_data->hash = FNV_prime * (uint64_t)fs.indicies[feature];

--- a/vowpalwabbit/lrq.cc
+++ b/vowpalwabbit/lrq.cc
@@ -112,7 +112,7 @@ void predict_or_learn(LRQstate& lrq, base_learner& base, example& ec)
                       if (all.audit || all.hash_inv)
                         { std::stringstream new_feature_buffer;
                           new_feature_buffer << right << '^'
-                                             << right_fs.space_names[rfn].second << '^'
+                                             << right_fs.space_names[rfn].get()->second << '^'
                                              << n;
 
 #ifdef _WIN32
@@ -122,7 +122,7 @@ void predict_or_learn(LRQstate& lrq, base_learner& base, example& ec)
                           char* new_space = strdup("lrq");
                           char* new_feature = strdup(new_feature_buffer.str().c_str());
 #endif
-                          right_fs.space_names.push_back(audit_strings(new_space,new_feature));
+                          right_fs.space_names.push_back(audit_strings_ptr(new audit_strings(new_space,new_feature)));
                         }
                     }
                 }

--- a/vowpalwabbit/lrqfa.cc
+++ b/vowpalwabbit/lrqfa.cc
@@ -74,7 +74,7 @@ void predict_or_learn(LRQFAstate& lrq, base_learner& base, example& ec)
                     if (all.audit || all.hash_inv)
                       { std::stringstream new_feature_buffer;
                         new_feature_buffer << right << '^'
-                                           << rfs.space_names[rfn].second << '^'
+                                           << rfs.space_names[rfn].get()->second << '^'
                                            << n;
 #ifdef _WIN32
                         char* new_space = _strdup("lrqfa");
@@ -83,7 +83,7 @@ void predict_or_learn(LRQFAstate& lrq, base_learner& base, example& ec)
                         char* new_space = strdup("lrqfa");
                         char* new_feature = strdup(new_feature_buffer.str().c_str());
 #endif
-                        rfs.space_names.push_back(audit_strings(new_space,new_feature));
+                        rfs.space_names.push_back(audit_strings_ptr(new audit_strings(new_space,new_feature)));
                       }
                   }
               }
@@ -114,7 +114,7 @@ void predict_or_learn(LRQFAstate& lrq, base_learner& base, example& ec)
 	  if (all.audit || all.hash_inv)
 	  {
 		for (size_t j = lrq.orig_size[right]; j < rfs.space_names.size(); ++j)
-		  rfs.space_names[j].~audit_strings(); 
+          rfs.space_names[j].~audit_strings_ptr();
 
           rfs.space_names.end = rfs.space_names.begin + lrq.orig_size[right];
         }

--- a/vowpalwabbit/parse_example.cc
+++ b/vowpalwabbit/parse_example.cc
@@ -102,7 +102,7 @@ public:
           { v_array<char> feature_v = v_init<char>();
             push_many(feature_v, feature_name.begin, feature_name.end - feature_name.begin);
             feature_v.push_back('\0');
-	    fs.space_names.push_back(audit_strings(base, feature_v.begin));
+            fs.space_names.push_back(audit_strings_ptr(new audit_strings(base, feature_v.begin)));
           }
         if ((affix_features[index] > 0) && (feature_name.end != feature_name.begin))
           {
@@ -130,7 +130,7 @@ public:
                      affix_v.push_back('=');
                      push_many(affix_v, affix_name.begin, affix_name.end - affix_name.begin);
                      affix_v.push_back('\0');
-		     fs.space_names.push_back(audit_strings("affix",affix_v.begin));
+                     fs.space_names.push_back(audit_strings_ptr(new audit_strings("affix",affix_v.begin)));
                    }
                  affix >>= 4;
               }
@@ -160,7 +160,7 @@ public:
           if (index != ' ') { spelling_v.push_back(index); spelling_v.push_back('_'); }
           push_many(spelling_v, spelling_ss.begin, spelling_ss.end - spelling_ss.begin);
           spelling_v.push_back('\0');
-          spell_fs.space_names.push_back(audit_strings("spelling",spelling_v.begin));
+          spell_fs.space_names.push_back(audit_strings_ptr(new audit_strings("spelling",spelling_v.begin)));
         }
       }
       if (namespace_dictionaries[index].size() > 0)
@@ -182,7 +182,7 @@ public:
                     ss << index << '_';
                     for (char* fc=feature_name.begin; fc!=feature_name.end; ++fc) ss << *fc;
                     ss << '=' << id;
-                    dict_fs.space_names.push_back(audit_strings("dictionary", ss.str()));
+                    dict_fs.space_names.push_back(audit_strings_ptr(new audit_strings("dictionary", ss.str())));
                   }
             }
         }

--- a/vowpalwabbit/parser.cc
+++ b/vowpalwabbit/parser.cc
@@ -613,12 +613,12 @@ void addgrams(vw& all, size_t ngram, size_t skip_gram, features& fs,
 
         fs.push_back(1.,new_index);
         if (fs.space_names.size() > 0)
-          { string feature_name(fs.space_names[i].second);
+          { string feature_name(fs.space_names[i].get()->second);
             for (size_t n = 1; n < gram_mask.size(); n++)
               { feature_name += string("^");
-                feature_name += string(fs.space_names[i+gram_mask[n]].second);
+                feature_name += string(fs.space_names[i+gram_mask[n]].get()->second);
               }
-			fs.space_names.push_back(audit_strings(fs.space_names[i].first, feature_name));
+            fs.space_names.push_back(audit_strings_ptr(new audit_strings(fs.space_names[i].get()->first, feature_name)));
           }
       }
   }
@@ -793,7 +793,7 @@ void add_constant_feature(vw& vw, example*ec)
   ec->feature_space[constant_namespace].push_back(1,constant);
   ec->total_sum_feat_sq++;
   ec->num_features++;
-  if (vw.audit || vw.hash_inv) ec->feature_space[constant_namespace].space_names.push_back(audit_strings("","Constant"));
+  if (vw.audit || vw.hash_inv) ec->feature_space[constant_namespace].space_names.push_back(audit_strings_ptr(new audit_strings("","Constant")));
 }
 
 void add_label(example* ec, float label, float weight, float base)

--- a/vowpalwabbit/search.cc
+++ b/vowpalwabbit/search.cc
@@ -458,7 +458,7 @@ void add_new_feature(search_private& priv, float val, uint64_t idx)
   {
     stringstream temp;
 	temp << "fid=" << ((idx & mask) >> ss) << "_" << priv.dat_new_feature_audit_ss.str();
-	fs.space_names.push_back(audit_strings(*priv.dat_new_feature_feature_space, temp.str()));
+    fs.space_names.push_back(audit_strings_ptr(new audit_strings(*priv.dat_new_feature_feature_space, temp.str())));
   }
 }
 

